File: Serialize-and-Deserialize-Binary-Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null ) return "N#";
        StringBuilder sb = new StringBuilder();
        Deque<TreeNode> dq = new LinkedList<>();
        dq.addFirst(root);
        while (! dq.isEmpty() ) {
            TreeNode t = dq.pollFirst();
            if (t==null) {
                sb.append("N#");
                continue;
                }
            sb.append(t.val+"#");
            dq.addLast(t.left);
            dq.addLast(t.right);
        }
        return sb.toString();

        
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) { 
        List<TreeNode> dq = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        char[] str = data.toCharArray();
        for(char s : str ){
            if(s == '#') { 
                String curr = sb.toString();
                if (curr.equals("N")) {
                    dq.add(null);

                }
                else {
                int temp = Integer.parseInt(curr);
                dq.add(new TreeNode(temp));
                }
                sb.setLength(0);
                continue;
            }
            sb.append(s);   
        }

        int parent =0;
        int child =1;
        while ( parent < dq.size() && child<dq.size()  )  { 
            if(dq.get(parent) == null ) {
                parent++;
                continue;
            }
            dq.get(parent).left = dq.get(child++); 
            dq.get(parent).right = dq.get(child++);
            parent++;
            

        }
        return dq.get(0);
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));

File: Serialize-and-Deserialize-Binary-Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        q = deque([root])
        res = ""
        while q:
            t=q.popleft()
            if t is None:
                res+= "N#"
            else:
                res+= str(t.val)+"#"
                q.append(t.left)
                q.append(t.right)
        return res
        
        



    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        def helper(d):
            l = deque()
            index= 0
             
            while index < len(d):
                b = ""
                while d[index]!="#":
                    b+= d[index]
                    index+=1
                if b =="N":
                    l.append(None)
                else:
                    l.append(TreeNode(int(b)))
                index += 1
            return l    
        def bfs(li):
            q = deque(li)
            root = q[0]
            index = 1
            curr = 0
            while q and index < len(q) and curr<len(q) : 
                t = q[curr]
                curr+=1
                if t is None:
                    continue
                
                if q[index]:
                    t.left = q[index]
                index+=1
                if q[index]:
                    t.right = q[index]
                index+=1
            return root
        return bfs(helper(data))





           


            

            
            

            
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))